### PR TITLE
Created handlers, requests structures and validated requests data

### PR DESCRIPTION
**Pull request by cyberdr0id**

### CODE REVIEW

File: pkg/hash/hash.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+func HashPassword(password string) (string, error) {
+	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 14)
+	return string(bytes), err
+}
+
+// CheckPasswordHash comparing password with his hash.
+func CheckPassowrdHash(password, hash string) bool {
+	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
+	return err == nil
+}
```
Comment (papahmsolo) [inline pkg/hash/hash.go:16]: No need to declare err here

File: internal/repository/database.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		config.Host, config.Port, config.User, config.DatabaseName, config.Password, config.SSLMode)
+
+	db, err := sql.Open("postgres", psqlconn)
+	if err != nil {
+		return nil, errSQLConnection
+	}
+
+	if err = db.Ping(); err != nil {
+		return nil, errPingDatabase
+	}
```
Comment (papahmsolo) [inline internal/repository/database.go:42]: - Can we wrap these errors giving some more context?
- Is there any particular reason we want to declare errors as `errSQLConnection`? Also note it's unexported

File: internal/handler/handlers.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	isValidSurname, _ := regexp.MatchString("(^[A-Za-zА-Яа-я]{2,16})?([ ]{0,1})([A-Za-zА-Яа-я]{2,16})?", r.candidateSurname)
+	if !isValidSurname {
+		return errInvalidName
+	}
+
+	return nil
+}
+
+// CheckDataCorrectness validates data after login/signup
+func (r *AuthRequest) CheckDataCorrectness() error {
```
Comment (papahmsolo) [inline internal/handler/handlers.go:192]: We're actually _validating_ the request, can we give it more descriptive name?

File: docs/swagger.yml [hunk 4]
```diff
@@ -47,20 +47,17 @@ paths:
...
+                type: integer
               example:
                 id: 256
-        400:
+        401:
           $ref: "#/components/responses/BadRequest"
         409:
           $ref: "#/components/responses/Conflict"
         500:
           $ref: "#/components/responses/InternalServerError"
```
Comment (papahmsolo) [inline docs/swagger.yml:59]: 401 is not a bad request, it's unauthenticated

File: internal/handler/handlers.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+// SignUp registers user.
+func SignUp(rw http.ResponseWriter, r *http.Request) {
+	var request AuthRequest
+
+	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
+		http.Error(rw, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (papahmsolo) [inline internal/handler/handlers.go:82]: Is it a server or user error?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
👀

File: internal/repository/database.go (hunk 2 above)
Comment (cyberdr0id) [reply internal/repository/database.go:42]: I decided to remove custom errors, because they have the same content as errors returned by methods

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good

File: internal/handler/handlers.go [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+	errParameterRequired = errors.New("input parameter is required")
+
+	// errInvalidFile presents an error when user load file with invalid name or wrong extension.
+	errInvalidFile = errors.New("invalid format or name of input file")
+
+	// errInvalidName presents an error when user send candidate with invalid name/surname.
+	errInvalidCandidateData = errors.New("input name didn't match to the desired format")
+
+	// errInvalidName presents an error when user try to login with wrong password.
+	errWrongPassword = errors.New("wrong password for inputed user")
```
Comment (papahmsolo) [inline internal/handler/handlers.go:27]: - common error "invalid param"

File: internal/handler/handlers.go [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+type CandidateSendingRequest struct {
+	FileName         string
+	CandidateName    string
+	CandidateSurname string
+}
+
+// UserRequests type presents structure which contains all user requests.
+type UserRequests struct {
+	Requests []repository.Request `json:"requests"`
+}
```
Comment (papahmsolo) [inline internal/handler/handlers.go:54]: - response?

File: internal/handler/handlers.go [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	if err := request.ValidateSignUpRequest(); err != nil {
+		http.Error(rw, err.Error(), http.StatusBadRequest)
+		return
+	}
+	pass, _ := hash.HashPassword(request.Password)
```
Comment (papahmsolo) [inline internal/handler/handlers.go:97]: - handle error

File: internal/handler/handlers.go [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+	id, err := s.Repo.CreateUser(request.Name, pass)
```
Comment (papahmsolo) [inline internal/handler/handlers.go:98]: service.CreateUser()

File: internal/handler/handlers.go [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+	var request LogInRequest
+
+	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
+		http.Error(rw, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer r.Body.Close()
+
+	user, err := s.Repo.GetUser(request.Name)
+	if errors.Is(err, repository.ErrNoUser) {
```
Comment (papahmsolo) [inline internal/handler/handlers.go:123]: service error user already exists

Comment (papahmsolo) [reply internal/handler/handlers.go:123]: handler interacts with service layer

File: internal/handler/handlers.go [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+// UpdateRequest updated status of request by id.
+func (s *Server) UpdateRequest(rw http.ResponseWriter, r *http.Request) {
+	rw.Header().Set("Content-Type", "application/json")
+
+	state := r.URL.Query().Get("state")
+
+	stateExp := "^([Aa]ccepted|[Rr]ejected|[Ss]ubmitted)$"
+	ok, err := regexp.MatchString(stateExp, state)
+	if !ok {
```
Comment (papahmsolo) [inline internal/handler/handlers.go:251]: validation function

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**